File: Maximize Happiness of Selected Children/Maximize Happiness of Selected Children.cpp
// Solution for Maximize Happiness of Selected Children in CPP

class Solution {
public:
    long long maximumHappinessSum(vector<int>& happiness, int k) {
        int n = happiness.size();

        // Sort happiness in descending order so we always pick the child with highest happiness
        sort(happiness.begin(), happiness.end(), greater<int>());

        long long totalHappiness = 0;

        // Select k children
        for (int turn = 0; turn < k; turn++) {
            // Happiness decreases by 1 for each previously selected child
            long long currentHappiness = max(happiness[turn] - turn, 0);
            totalHappiness += currentHappiness;
        }

        return totalHappiness;
    }
};

File: Maximize Happiness of Selected Children/Maximize Happiness of Selected Children.py
# Solution for Maximize Happiness of Selected Children in PY

class Solution:
    def maximumHappinessSum(self, happiness: List[int], k: int) -> int:

        # Sort happiness in descending order
        happiness.sort(reverse=True)

        total_happiness = 0

        # Select k children
        for turn in range(k):
            # Happiness decreases by 1 for each previously selected child
            current_happiness = max(happiness[turn] - turn, 0)
            total_happiness += current_happiness

        return total_happiness
